File: App/ms-zonas-espacios/src/main/java/ec/edu/espe/zonas/tenant/TenantAliases.java
package ec.edu.espe.zonas.tenant;

import jakarta.servlet.http.HttpServletRequest;

import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

/**
 * Defensa en profundidad frente a la invasion de tenant: un usuario del tenant
 * A que intenta leer datos del tenant B reenviando el identificador ajeno en
 * una cabecera propia (X-Tenant: empresa-b) o en la URL (?tenant=empresa-b).
 *
 * Kong ya rechaza esas peticiones en el borde comparandolas contra el claim
 * firmado, pero el microservicio no debe depender de ello: si alguien alcanza
 * el Service de Kubernetes directamente, sin pasar por el gateway, la misma
 * comprobacion tiene que aplicarse aqui.
 *
 * El tenant efectivo es el que ya validaron TenantContextFilter (cabecera
 * canonica) y JwtAuthenticationFilter (claim del token). Esta clase solo
 * comprueba que ningun alias adicional contradiga ese valor.
 */
public final class TenantAliases {

    /** Nombres de cabecera con los que se suele intentar suplantar el tenant. */
    private static final List<String> CABECERAS = List.of(
            "X-Tenant-ID", "X-Tenant", "X-TenantId", "X-Tenant-Name",
            "Tenant", "Tenant-Id", "TenantId",
            "X-Empresa", "Empresa", "X-Empresa-Id");

    /** Parametros de URL equivalentes (aqui si importan las mayusculas). */
    private static final List<String> PARAMETROS = List.of(
            "tenant_id", "tenant", "tenantId", "tenantid", "tenant-id",
            "empresa", "empresa_id", "empresaId", "idEmpresa");

    private TenantAliases() {
    }

    /**
     * @param tenantEfectivo tenant ya normalizado y autorizado para la peticion
     * @return descripcion del primer alias que contradice a {@code tenantEfectivo},
     *         o {@code null} si todos coinciden (o no se envio ninguno)
     */
    public static String discrepancia(HttpServletRequest request, String tenantEfectivo) {
        for (String cabecera : CABECERAS) {
            String valor = request.getHeader(cabecera);
            if (discrepa(valor, tenantEfectivo)) {
                return "cabecera " + cabecera + "=" + valor;
            }
        }
        String query = request.getQueryString();
        for (String parametro : PARAMETROS) {
            String valor = parametroDeUrl(query, parametro);
            if (discrepa(valor, tenantEfectivo)) {
                return "parametro " + parametro + "=" + valor;
            }
        }
        return null;
    }

    private static boolean discrepa(String valor, String tenantEfectivo) {
        return valor != null
                && !valor.isBlank()
                && !valor.trim().toLowerCase().equals(tenantEfectivo);
    }

    /**
     * Lee un parametro directamente de la query string.
     *
     * No se usa request.getParameter() a proposito: ante un POST de tipo
     * application/x-www-form-urlencoded ese metodo consume el cuerpo de la
     * peticion y el controlador lo recibiria vacio.
     */
    private static String parametroDeUrl(String query, String nombre) {
        if (query == null || query.isEmpty()) {
            return null;
        }
        for (String par : query.split("&")) {
            int igual = par.indexOf('=');
            String clave = igual >= 0 ? par.substring(0, igual) : par;
            if (clave.equals(nombre)) {
                String valor = igual >= 0 ? par.substring(igual + 1) : "";
                return URLDecoder.decode(valor, StandardCharsets.UTF_8);
            }
        }
        return null;
    }
}

File: App/ms-audith/src/common/tenant-aliases.ts
/**
 * Defensa en profundidad frente a la invasion de tenant: un usuario del tenant
 * A que intenta leer datos del tenant B reenviando el identificador ajeno en
 * una cabecera propia (X-Tenant: empresa-b) o en la URL (?tenant=empresa-b).
 *
 * Kong ya rechaza esas peticiones en el borde comparandolas contra el claim
 * firmado del JWT, pero el microservicio no debe depender de ello: si alguien
 * alcanza el Service de Kubernetes directamente, sin pasar por el gateway, la
 * misma comprobacion tiene que aplicarse aqui.
 */

/** Express normaliza los nombres de cabecera a minusculas. */
const CABECERAS = [
  'x-tenant-id',
  'x-tenant',
  'x-tenantid',
  'x-tenant-name',
  'tenant',
  'tenant-id',
  'tenantid',
  'x-empresa',
  'empresa',
  'x-empresa-id',
];

/** Parametros de URL equivalentes (aqui si importan las mayusculas). */
const PARAMETROS = [
  'tenant_id',
  'tenant',
  'tenantId',
  'tenantid',
  'tenant-id',
  'empresa',
  'empresa_id',
  'empresaId',
  'idEmpresa',
];

/** Un alias repetido llega como arreglo; solo interesa el primer valor. */
function primerValor(valor: unknown): string | undefined {
  if (Array.isArray(valor)) {
    return typeof valor[0] === 'string' ? valor[0] : undefined;
  }
  return typeof valor === 'string' ? valor : undefined;
}

function discrepa(valor: string | undefined, tenantEfectivo: string): boolean {
  return (
    valor !== undefined &&
    valor.trim() !== '' &&
    valor.trim().toLowerCase() !== tenantEfectivo
  );
}

/**
 * @param tenantEfectivo tenant ya normalizado y autorizado para la peticion
 * @returns descripcion del primer alias que contradice a `tenantEfectivo`, o
 *          `null` si todos coinciden (o no se envio ninguno)
 */
export function discrepanciaDeTenant(
  request: any,
  tenantEfectivo: string,
): string | null {
  for (const nombre of CABECERAS) {
    const valor = primerValor(request?.headers?.[nombre]);
    if (discrepa(valor, tenantEfectivo)) {
      return `cabecera ${nombre}=${valor}`;
    }
  }
  for (const nombre of PARAMETROS) {
    const valor = primerValor(request?.query?.[nombre]);
    if (discrepa(valor, tenantEfectivo)) {
      return `parametro ${nombre}=${valor}`;
    }
  }
  return null;
}
